runtime/compiler/control/CompilationThread.cpp

```diff
@@ -7139,6 +7162,42 @@ TR::CompilationInfoPerThreadBase::exitPerClientAllocationRegion()
       _compiler->switchToGlobalMemory();
    _perClientPersistentMemory = NULL;
    }
+
+void
+TR::CompilationInfoPerThreadBase::downgradeLocalCompilationIfLowPhysicalMemory(TR_MethodToBeCompiled *entry, bool forcedLocal)
+   {
+   TR_ASSERT_FATAL(_compInfo.getPersistentInfo()->getRemoteCompilationMode() == JITServer::CLIENT,
+                   "Must be called on JITServer client");
+
+   J9Method *method = entry->getMethodDetails().getMethod();
+   if (forcedLocal &&
+       !TR::Options::getCmdLineOptions()->getOption(TR_DisableJITServerBufferedExpensiveCompilations) &&
+       TR::Options::getCmdLineOptions()->allowRecompilation() &&
+       !TR::CompilationInfo::isCompiled(method) &&  // do not downgrade recompilations
+       entry->_optimizationPlan->getOptLevel() >= warm)
+      {
+      // Downgrade a forced local compilation, if the available client memory is low.
+      bool incomplete;
+      uint64_t freePhysicalMemorySizeB = _compInfo.computeAndCacheFreePhysicalMemory(incomplete);
+      int32_t numActiveThreads = _compInfo.getNumCompThreadsActive();
+      if (freePhysicalMemorySizeB != OMRPORT_MEMINFO_NOT_AVAILABLE &&
+         freePhysicalMemorySizeB <= (uint64_t)TR::Options::getSafeReservePhysicalMemoryValue() + (numActiveThreads + 4) * TR::Options::getScratchSpaceLowerBound())
```

Comment:
Could we please measure the average/maximum amount of scratch memory at cold for something like AcmeAir to make sure the value here is enough?